File: Projeto2020_API/Validators/FuncionarioValidator.cs
using FluentValidation;
using Dominio.Dtos;

namespace Projeto2025_API.Validators
{
    public class FuncionarioValidator : AbstractValidator<FuncionarioDTO>
    {
        public FuncionarioValidator()
        {
            // Validações baseadas no formulário GerenciarFuncionarios.tsx
            
            RuleFor(x => x.Nome)
                .NotEmpty().WithMessage("Nome é obrigatório")
                .Length(2, 100).WithMessage("Nome deve ter entre 2 e 100 caracteres")
                .Matches(@"^[a-zA-ZÀ-ÿ\s]+$").WithMessage("Nome deve conter apenas letras e espaços");

            RuleFor(x => x.Email)
                .NotEmpty().WithMessage("Email é obrigatório")
                .EmailAddress().WithMessage("Email deve ser válido")
                .MaximumLength(255).WithMessage("Email deve ter no máximo 255 caracteres");

            RuleFor(x => x.Telefone)
                .Must(ValidarTelefone).WithMessage("Telefone deve estar no formato (00) 00000-0000")
                .When(x => !string.IsNullOrEmpty(x.Telefone));

            RuleFor(x => x.Senha)
                .NotEmpty().WithMessage("Senha é obrigatória")
                .MinimumLength(6).WithMessage("A senha deve ter pelo menos 6 caracteres")
                .When(x => x.Id == 0); // Apenas para novos funcionários

            RuleFor(x => x.Cargo)
                .NotEmpty().WithMessage("Cargo é obrigatório")
                .MaximumLength(50).WithMessage("Cargo deve ter no máximo 50 caracteres");

            RuleFor(x => x.Salario)
                .NotEmpty().WithMessage("Salário é obrigatório")
                .GreaterThan(0).WithMessage("Salário deve ser maior que 0");

            RuleFor(x => x.DataAdmissao)
                .NotEmpty().WithMessage("Data de admissão é obrigatória")
                .LessThanOrEqualTo(DateTime.Today).WithMessage("Data de admissão não pode ser futura");

            RuleFor(x => x.DataDemissao)
                .GreaterThan(x => x.DataAdmissao).WithMessage("Data de demissão deve ser posterior à admissão")
                .When(x => x.DataDemissao.HasValue);
        }

        private bool ValidarTelefone(string telefone)
        {
            if (string.IsNullOrEmpty(telefone)) return true;
            
            var telefoneRegex = @"^\(\d{2}\)\s\d{4,5}-\d{4}$";
            return System.Text.RegularExpressions.Regex.IsMatch(telefone, telefoneRegex);
        }
    }
}

File: frontend-yeti/src/validators/FuncionarioValidator.ts
export class FuncionarioValidator {
    static validateNome(nome: string): string {
        if (!nome || !nome.trim()) return 'Nome é obrigatório';
        if (nome.trim().length < 2) return 'Nome deve ter pelo menos 2 caracteres';
        if (nome.trim().length > 100) return 'Nome deve ter no máximo 100 caracteres';
        return '';
    }

    static validateEmail(email: string): string {
        if (!email || !email.trim()) return 'Email é obrigatório';
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        if (!emailRegex.test(email.trim())) return 'Email deve ter um formato válido';
        if (email.trim().length > 100) return 'Email deve ter no máximo 100 caracteres';
        return '';
    }

    static validateTelefone(telefone: string): string {
        if (!telefone || !telefone.trim()) return 'Telefone é obrigatório';
        const telefoneLimpo = telefone.replace(/\D/g, '');
        if (telefoneLimpo.length < 10) return 'Telefone deve ter pelo menos 10 dígitos';
        if (telefoneLimpo.length > 11) return 'Telefone deve ter no máximo 11 dígitos';
        return '';
    }

    static validateSenha(senha: string, isEdit: boolean = false): string {
        if (!isEdit && (!senha || !senha.trim())) return 'Senha é obrigatória';
        if (senha && senha.trim().length < 6) return 'Senha deve ter pelo menos 6 caracteres';
        if (senha && senha.trim().length > 50) return 'Senha deve ter no máximo 50 caracteres';
        return '';
    }

    static validateCargo(cargo: string): string {
        if (!cargo || !cargo.trim()) return 'Cargo é obrigatório';
        return '';
    }

    static validateSalario(salario: number): string {
        if (!salario || salario <= 0) return 'Salário deve ser maior que zero';
        if (salario > 999999.99) return 'Salário deve ser menor que R$ 999.999,99';
        return '';
    }

    static validateDataAdmissao(dataAdmissao: string): string {
        if (!dataAdmissao || !dataAdmissao.trim()) return 'Data de admissão é obrigatória';
        
        const data = new Date(dataAdmissao);
        const hoje = new Date();
        hoje.setHours(0, 0, 0, 0);
        
        if (isNaN(data.getTime())) return 'Data de admissão inválida';
        if (data > hoje) return 'Data de admissão não pode ser futura';
        
        // Verificar se não é muito antiga (mais de 50 anos)
        const cinquentaAnosAtras = new Date();
        cinquentaAnosAtras.setFullYear(hoje.getFullYear() - 50);
        if (data < cinquentaAnosAtras) return 'Data de admissão muito antiga';
        
        return '';
    }

    static validateDataDemissao(dataDemissao: string, dataAdmissao: string): string {
        if (!dataDemissao || !dataDemissao.trim()) return ''; // Opcional
        
        const dataDem = new Date(dataDemissao);
        const dataAdm = new Date(dataAdmissao);
        
        if (isNaN(dataDem.getTime())) return 'Data de demissão inválida';
        if (dataDem < dataAdm) return 'Data de demissão deve ser posterior à data de admissão';
        
        return '';
    }

    static validateForm(formData: any, isEdit: boolean = false): Record<string, string> {
        const errors: Record<string, string> = {};

        errors.nome = this.validateNome(formData.nome);
        errors.email = this.validateEmail(formData.email);
        errors.telefone = this.validateTelefone(formData.telefone);
        errors.senha = this.validateSenha(formData.senha, isEdit);
        errors.cargo = this.validateCargo(formData.cargo);
        errors.salario = this.validateSalario(formData.salario);
        errors.dataAdmissao = this.validateDataAdmissao(formData.dataAdmissao);
        errors.dataDemissao = this.validateDataDemissao(formData.dataDemissao, formData.dataAdmissao);

        // Remover erros vazios
        Object.keys(errors).forEach(key => {
            if (!errors[key]) {
                delete errors[key];
            }
        });

        return errors;
    }
}
